compiler/src/dotty/tools/dotc/transform/Splicing.scala

```diff
@@ -0,0 +1,394 @@
+          if isCaptured(lhs.symbol) then transformSplicedAssign(tree)
+          else super.transform(tree)
+        case Apply(fn, args) if fn.symbol == defn.QuotedRuntime_exprNestedSplice =>
+          val newArgs = args.mapConserve(arg => transform(arg)(using spliceContext))
+          cpy.Apply(tree)(fn, newArgs)
+        case Apply(sel @ Select(app @ Apply(fn, args),nme.apply), quotesArgs)
+        if fn.symbol == defn.QuotedRuntime_exprQuote =>
+          args match
+            case List(tree: RefTree) if isCaptured(tree.symbol) =>
+              capturedTerm(tree)
+            case _ =>
+              val newArgs = withCurrentQuote(quotesArgs.head) {
+                if level > 1 then args.mapConserve(arg => transform(arg)(using quoteContext))
+                else args.mapConserve(arg => transformLevel0QuoteContent(arg)(using quoteContext))
+              }
+              cpy.Apply(tree)(cpy.Select(sel)(cpy.Apply(app)(fn, newArgs), nme.apply), quotesArgs)
+        case Apply(TypeApply(_, List(tpt: Ident)), List(quotes))
+        if tree.symbol == defn.QuotedTypeModule_of && isCaptured(tpt.symbol) =>
+          ref(capturedType(tpt))
+        case CapturedApplication(fn, argss) =>
+          transformCapturedApplication(tree, fn, argss)
+        case _ =>
+          super.transform(tree)
+
+    private def transformLevel0QuoteContent(tree: Tree)(using Context): Tree =
+      // transform and collect new healed types
+      val old = healedTypes
+      healedTypes = new PCPCheckAndHeal.QuoteTypeTags(tree.span)
+      val tree1 = transform(tree)
+      val newHealedTypes = healedTypes.getTypeTags
+      healedTypes = old
+      // add new healed types to the current, merge with existing healed types if necessary
+      if newHealedTypes.isEmpty then tree1
+      else tree1 match
+        case Block(stats @ (x :: _), expr) if x.symbol.hasAnnotation(defn.QuotedRuntime_SplicedTypeAnnot) =>
+          Block(newHealedTypes ::: stats, expr)
+        case _ =>
+          Block(newHealedTypes, tree1)
+
+    class ArgsClause(val args: List[Tree]):
+      def isTerm: Boolean = args.isEmpty || args.head.isTerm
+
+    private object CapturedApplication {
+
+      /** Matches and application `f(...)` (possibly with several argument clauses) where `f` is captured */
+      def unapply(tree: Tree)(using Context): Option[(RefTree, List[ArgsClause])] = tree match
+        case GenericApply(fn: RefTree, args) if isCaptured(fn.symbol) =>
+          Some((fn, ArgsClause(args) :: Nil))
+        case GenericApply(CapturedApplication(fn, argss), args) =>
+          Some((fn, argss :+ ArgsClause(args)))
+        case _ =>
+          None
+    }
+
+    private def containsCapturedType(tpe: Type)(using Context): Boolean =
+      tpe match
+        case tpe @ TypeRef(prefix, _) => isCaptured(tpe.symbol) || containsCapturedType(prefix)
+        case tpe @ TermRef(prefix, _) => isCaptured(tpe.symbol) || containsCapturedType(prefix)
```

Comment:
I would do:
```suggestion
        case tpe: NamedType => isCaptured(tpe.symbol) || containsCapturedType(tpe.prefix)
```
Also I don't know the context here, but is this really complete? I would expect something like `tpe.existsPart(isCaptured)` to deal with aliases, type lambdas, etc (also existsPart takes a stopAt argument that can be used to avoid traversing static prefixes which I assume would make sense here).

Reply:
Indeed, `existsPart` seems to be what I need to use here.